src/KafkaFlow.IntegrationTests/OpenTelemetryTests.cs

```diff
@@ -0,0 +1,174 @@
+﻿namespace KafkaFlow.IntegrationTests
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using AutoFixture;
+    using global::OpenTelemetry;
+    using global::OpenTelemetry.Trace;
+    using KafkaFlow.Compressor;
+    using KafkaFlow.Compressor.Gzip;
+    using KafkaFlow.Configuration;
+    using KafkaFlow.IntegrationTests.Core;
+    using KafkaFlow.IntegrationTests.Core.Handlers;
+    using KafkaFlow.IntegrationTests.Core.Middlewares;
+    using KafkaFlow.IntegrationTests.Core.Producers;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.Hosting;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class OpenTelemetryTests
+    {
+        private readonly Fixture fixture = new();
+
+        private List<Activity> exportedItems;
+
+        private IServiceProvider provider;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            this.exportedItems = new List<Activity>();
+            this.provider = this.GetServiceProvider();
+            MessageStorage.Clear();
+        }
+
+        [TestMethod]
+        public async Task AddOpenTelemetry_ProducingAndConsumingOneMessage_TraceAndSpansAreCreatedCorrectly()
+        {
+            // Arrange
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddSource("KafkaFlow.OpenTelemetry")
+            .AddInMemoryExporter(this.exportedItems)
+            .Build();
+
+            var producer = this.provider.GetRequiredService<IMessageProducer<GzipProducer>>();
+            var message = this.fixture.Create<byte[]>();
+
+            // Act
+            await producer.ProduceAsync(null, message);
+
+            // Assert
+            await Task.Delay(8000).ConfigureAwait(false);
```

Comment:
Replace sleep's with retry pattern please.

Reply:
The same as https://github.com/Farfetch/kafkaflow/pull/451#discussion_r1360801150.